software/apps/cholesky_schedule_q32s/main.c

```diff
@@ -0,0 +1,247 @@
+  mempool_barrier_init(core_id);
+  /* Initialize input matrices */
+  init_matrix(A_matrix, N, N, -156, 427, -219, core_id);
+  init_matrix_zeros(AT_matrix, N, N, core_id);
+  init_matrix_zeros(M_matrix, N, N, core_id);
+  init_matrix_zeros(In_matrix, N, N_BANKS, core_id);
+  /* Initialize output matrices */
+  init_matrix_zeros(LL_matrix, N_ROW * N, N_BANKS, core_id);
+  init_matrix_zeros(LR_matrix, N_ROW * N, N_BANKS, core_id);
+  mempool_barrier(num_cores);
+  /* Create positive definite matrix */
+  if (core_id == 0) {
+    transpose(A_matrix, AT_matrix, N);
+    matrixmult(AT_matrix, A_matrix, M_matrix, N);
+    for (uint32_t idx_col = 0; idx_col < N_COL; idx_col++) {
+      for (uint32_t i = 0; i < N; i++) {
+        for (uint32_t j = 0; j < N; j++) {
+          In_matrix[idx_col * N + i * N_BANKS + j] = M_matrix[i * N + j];
+        }
+      }
+    }
+    printf("Done initialization.\n");
+  }
+  mempool_barrier(num_cores);
+  /* Benchmark */
+  uint32_t nPE = (N / 4);
+  /* The decomposition is finely-grained parallelized over multiple cores */
+  if (nPE > 1) {
+    if (core_id < N_COL * nPE) {
+      mempool_start_benchmark();
+      mempool_cholesky_q32p_fold(In_matrix, In_matrix, LL_matrix, LR_matrix, N,
+                                 N_ROW, N_COL);
+      mempool_stop_benchmark();
+    }
+    mempool_barrier(num_cores);
+  }
+  /* The decomposition is executed with a single-core. Each core gets a
+   * different input problem. This is the specific case of the 4x4 matrix. */
+  if (nPE == 1) {
+    if (core_id < N_COL * nPE) {
+      mempool_start_benchmark();
+      mempool_cholesky_q32s_schedule(In_matrix, LL_matrix, N, N_ROW, N_COL);
+      mempool_stop_benchmark();
+    }
+    mempool_barrier(num_cores);
+  }
+  return;
+}
+
+// Driver program
+void parallel_linearsolver() {
+  uint32_t core_id = mempool_get_core_id();
+  uint32_t num_cores = mempool_get_core_count();
+  mempool_barrier_init(core_id);
+  /* Initialize input matrices */
+  init_matrix(A_matrix, N, N, -156, 427, -219, core_id);
+  init_matrix_zeros(AT_matrix, N, N, core_id);
+  init_matrix_zeros(M_matrix, N, N, core_id);
+  init_matrix_zeros(In_matrix, N, N_BANKS, core_id);
+  init_matrix(In, 1, N_BANKS, -156, 427, -219, core_id);
+  /* Initialize output matrices */
+  init_matrix_zeros(LL_matrix, N_ROW * N, N_BANKS, core_id);
+  init_matrix_zeros(LR_matrix, N_ROW * N, N_BANKS, core_id);
+  mempool_barrier(num_cores);
+  /* Create positive definite matrix */
+  if (core_id == 0) {
+    transpose(A_matrix, AT_matrix, N);
+    matrixmult(AT_matrix, A_matrix, M_matrix, N);
+    for (uint32_t idx_col = 0; idx_col < N_COL; idx_col++) {
+      for (uint32_t i = 0; i < N; i++) {
+        for (uint32_t j = 0; j < N; j++) {
+          In_matrix[idx_col * N + i * N_BANKS + j] = M_matrix[i * N + j];
+        }
+      }
+    }
+    printf("Done initialization.\n");
+  }
+  mempool_barrier(num_cores);
+  /* Benchmark */
+  uint32_t nPE = (N / 4);
```

Comment:
Here there is a lot of duplicate code and indirection again. Can this be reduced somehow? I see that there is more going on in this file around the individual kernels, but the init setup is still the same for all benchmarks.

Reply:
Done